lib/internal/quic/core.js

```diff
@@ -0,0 +1,3341 @@
+
+    this.#maybeReady(0x2);
+  }
+
+  [kSocketReady]() {
+    this.#maybeReady(0x1);
+  }
+
+  // The QuicClientSession is ready for use only after
+  // (a) The QuicSocket has been bound and
+  // (b) The internal handle has been created
+  #maybeReady = function(flag) {
+    this.#ready |= flag;
+    if (this.ready)
+      process.nextTick(emit.bind(this, 'ready'));
+  };
+
+  get allowEarlyData() {
+    return this.#allowEarlyData;
+  }
+
+  get ready() {
+    return this.#ready === 0x3;
+  }
+
+  get handshakeStarted() {
+    return this.#handshakeStarted;
+  }
+
+  startHandshake() {
+    if (this.destroyed)
+      throw new ERR_QUICSESSION_DESTROYED('startHandshake');
+    if (this.#handshakeStarted)
+      return;
+    this.#handshakeStarted = true;
+    if (!this.ready) {
+      this.once('ready', () => this[kHandle].startHandshake());
+    } else {
+      this[kHandle].startHandshake();
+    }
+  }
+
+  get ephemeralKeyInfo() {
+    return this[kHandle] !== undefined ?
+      this[kHandle].getEphemeralKeyInfo() :
+      {};
+  }
+
+  #setSocketAfterBind = function(socket, callback) {
+    if (socket.destroyed) {
+      callback(new ERR_QUICSOCKET_DESTROYED('setSocket'));
+      return;
+    }
+
+    if (!this[kHandle].setSocket(socket[kHandle])) {
+      callback(new ERR_QUICCLIENTSESSION_FAILED_SETSOCKET());
+      return;
+    }
+
+    if (this.socket) {
+      this.socket[kRemoveSession](this);
+      this[kSetSocket](undefined);
+    }
+    socket[kAddSession](this);
+    this[kSetSocket](socket);
+
+    callback();
+  };
+
+  setSocket(socket, callback) {
+    if (!(socket instanceof QuicSocket))
+      throw new ERR_INVALID_ARG_TYPE('socket', 'QuicSocket', socket);
+
+    if (typeof callback !== 'function')
+      throw new ERR_INVALID_CALLBACK();
+
+    socket[kMaybeBind](() => this.#setSocketAfterBind(socket, callback));
+  }
+}
+
+function streamOnResume() {
+  if (!this.destroyed)
+    this[kHandle].readStart();
+}
+
+function streamOnPause() {
+  if (!this.destroyed /* && !this.pending */)
+    this[kHandle].readStop();
+}
+
+class QuicStream extends Duplex {
+  #closed = false;
+  #aborted = false;
+  #defaultEncoding = undefined;
+  #didRead = false;
+  #id = undefined;
+  #highWaterMark = undefined;
+  #push_id = undefined;
+  #resetCode = undefined;
+  #session = undefined;
+  #dataRateHistogram = undefined;
+  #dataSizeHistogram = undefined;
+  #dataAckHistogram = undefined;
+  #stats = undefined;
+
+  constructor(options, session, push_id) {
+    const {
+      highWaterMark,
+      defaultEncoding,
+    } = options;
+    super({
+      highWaterMark,
+      defaultEncoding,
+      allowHalfOpen: true,
+      decodeStrings: true,
+      emitClose: true,
+      autoDestroy: false,
+      captureRejections: true,
+    });
+    this.#highWaterMark = highWaterMark;
+    this.#defaultEncoding = defaultEncoding;
+    this.#session = session;
+    this.#push_id = push_id;
+    this._readableState.readingMore = true;
```

Comment:
This is a bit non-standard. A comment explaining?

Reply:
Will address in a subsequent PR